include/drivers/dai.h

```diff
@@ -0,0 +1,484 @@
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+	int type; /* SSP, DMIC, HDA, etc. */
+	int direction; /* Playback, capture */
+	int index; /* For SSPx to select correct timestamp register */
+	int dma_id; /* DMA instance id */
+	int dma_chan_index; /* Used DMA channel */
+	int dma_chan_count; /* Channels in single DMA */
+};
+
+struct dai_ts_data {
+	uint64_t walclk; /* Wall clock */
+	uint64_t sample; /* Sample count */
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * For internal use only, skip these in public documentation.
+ */
+__subsystem struct dai_driver_api {
+	int (*init)(const struct device *dev);
+	int (*deinit)(const struct device *dev);
+
+	int (*config_set)(const struct device *dev, const struct dai_config *cfg,
+			  const void * bespoke_cfg);
+	const struct dai_config *(*config_get)(const struct device *dev,
+					       enum dai_dir dir);
+
+	const struct dai_properties *(*get_properties)(const struct device *dev,
+						       enum dai_dir dir,
+						       int stream_id);
+
+	int (*trigger)(const struct device *dev, enum dai_dir dir,
+		       enum dai_trigger_cmd cmd);
+
+	int (*read)(const struct device *dev, void *buf, size_t size);
+	int (*write)(const struct device *dev, void *buf, size_t size);
+
+	/* timestamp ops - optional */
+	int (*ts_config)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_start)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_stop)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_get)(const struct device *dev, struct dai_ts_cfg *cfg,
+		      struct dai_ts_data *tsd);
+};
+/**
+ * @endcond
+ */
+
+static inline int dai_init(const struct device *dev)
+{
+	const struct dai_driver_api *api =
+		(const struct dai_driver_api *)dev->api;
+
+	return api->init(dev);
+}
+
+static inline int dai_deinit(const struct device *dev)
+{
+	const struct dai_driver_api *api =
+		(const struct dai_driver_api *)dev->api;
+
+	return api->deinit(dev);
+}
+
+/**
+ * @brief Configure operation of a host DAI controller.
+ *
+ * The dir parameter specifies if Transmit (TX) or Receive (RX) direction
+ * will be configured by data provided via cfg parameter.
+ *
+ * The function can be called in NOT_READY or READY state only. If executed
+ * successfully the function will change the interface state to READY.
+ *
+ * If the function is called with the parameter cfg->frame_clk_freq set to 0
+ * the interface state will be changed to NOT_READY.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cfg Pointer to the structure containing configuration parameters.
+ * @param bespoke_cfg Pointer to the structure containing bespoke config.
+ *
+ * @retval 0 If successful.
+ * @retval -EINVAL Invalid argument.
+ * @retval -ENOSYS DAI_DIR_BOTH value is not supported.
+ */
+static inline int dai_config_set(const struct device *dev,
+				 const struct dai_config *cfg,
+				 const void *bespoke_cfg)
```

Comment:
yes agreed. So the idea is that this would be more general than I2S -> we would have different bespoke parameters for different drivers. Now for example our existing dmic driver has huge parameter set and it can operate in 2 different modes: typed nested structs or binary blobs with direct register layout. So when having several different drivers like this, I see only option to have a union of these structs as parameter and it is not looking very nice...

So I'm wondering how is the validation specified to be done? Is there any examples that one could send like non-typed data through the API and the driver implementation would be responsible of the validation (because the driver type understands what kind of data to expect)?

Reply:
void * unfortunately means, even if you have a pointer to some device specific config like struct my_dai_cfg{}, and a size of that struct pass as a param here so you could copy it, you can't guarantee the struct and the driver don't expect more pointers inside that cfg, which is a no-no from reading the docs. You need to be able to verify and copy any parameters that are used for branching based on the docs here:

https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html#verification-function

And that makes sense, if you have branch statements in supervisor mode based on data pass in, you need to ensure the data is 1) valid at the time of the call, 2) doesn't change once the call is started. Neither is possible really with a void*.

Really though this *only*  matters if you care about ever supporting usermode syscalls to this particular API. You could simply say this is a supervisor only call, and not sure the __syscall macro. That doesn't prevent using __syscalls with other APIs here either, everything else seems perfectly valid.

Reply:
What the gist of the comment above boils down to, is void* won't preclude the API from being a syscall at some point but that there might need to be some inventive thinking around it. That's ok, marking this as resolved.